pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go

```diff
@@ -403,26 +403,85 @@ func (m panicNodeMutator) Generate(
 	rng *rand.Rand, plan *TestPlan, planner *testPlanner,
 ) []mutation {
 	var mutations []mutation
-	maxPanics := len(plan.upgrades)
-	numPanics := 1 + rng.Intn(maxPanics)
-
+	upgrades := randomUpgrades(rng, plan)
 	idx := newStepIndex(plan)
-	possiblePointsInTime := plan.
-		newStepSelector().
-		// We don't want to panic concurrently with other steps, and inserting before a concurrent step
-		// causes the step to run concurrently with that step, so we filter out any concurrent steps.
-		Filter(func(s *singleStep) bool {
-			return s.context.System.Stage >= InitUpgradeStage && !idx.IsConcurrent(s)
-		})
+	nodeList := planner.currentContext.System.Descriptor.Nodes
+
+	for _, upgrade := range upgrades {
+		possiblePointsInTime := upgrade.
+			// We don't want to panic concurrently with other steps, and inserting before a concurrent step
+			// causes the step to run concurrently with that step, so we filter out any concurrent steps.
+			Filter(func(s *singleStep) bool {
+				return s.context.System.Stage >= InitUpgradeStage && !idx.IsConcurrent(s)
+			})
 
-	for range numPanics {
-		nodeList := planner.currentContext.System.Descriptor.Nodes
 		targetNode := nodeList.SeededRandNode(rng)
-		addRandomly := possiblePointsInTime.
-			RandomStep(rng).
-			InsertBefore(panicNodeStep{planner.currentContext.System.Descriptor.Nodes[0], targetNode, planner.rt})
+		stepToPanic := possiblePointsInTime.RandomStep(rng)
+		hasInvalidConcurrentStep := false
+		concurrentStepMarked := false
+		var concurrentStep *singleStep
+
+		isIncompatibleStep := func(s *singleStep) bool {
+			// Restarting the system on a node while a different node is already dead can
+			// cause the cluster to lose quorum, and we cannot run restartWithNewBinaryStep
```

Comment:
I think we chatted about this but don't remember the conclusion. It's not that restarting the node causes the cluster to lose quorum, the node is already dead at that point. It's _failing to restart_ the node that could cause us to lose quorum.